### ISSUE TITLE
[Event request] Page "Permission Sets" (ID 9802) - OnBeforeImportPermissionSets

### ISSUE THREAD

**Issue by iaxmartinwagner** [microsoft/ALAppExtensions#22221]
Hello,

when importing permission sets it´s only possible to add or "merge" permissions to existing ones.
You do not have a change to delete existing permission when you are importing reduced permission sets.
For this case we need an event "OnBeforeImportPermissionSets" to check the import file if there are permission sets included in importfile which still exist, perform a user request "should existing permissions to be deleted before import starts", if yes deletes existing permissions for all permission sets in importfile and then go on with normal import.

The function for the event should called in action "ImportPermissionSets" and handled pattern should avoid a second user request.
```
            action(ImportPermissionSets)
            {
                ApplicationArea = Basic, Suite;
                Caption = 'Import Permission Sets';
                Enabled = CanManageUsersOnTenant;
                Image = Import;
                ToolTip = 'Import a file with permissions.';

                trigger OnAction()
                var
                    NSYSPermissionMgt: Codeunit "IAX NSYS Permission Mgt.";
                    PermissionSetBuffer: Record "Permission Set Buffer";
                    TempBlob: Codeunit "Temp Blob";
                    ImportPermissionSets: XmlPort "Import Permission Sets";
#if not CLEAN21
                    ImportTenantPermissionSets: XmlPort "Import Tenant Permission Sets";
#endif
                    FileName: Text;
                    InStream: InStream;
                    OutStream: OutStream;
                    UpdateExistingPermissions: Boolean;
                    Handeld: Boolean;
                begin
                    UploadIntoStream('Import', '', '', FileName, InStream);
                    TempBlob.CreateOutStream(OutStream);
                    CopyStream(OutStream, InStream);

                    TempBlob.CreateInStream(InStream);
                    OnBeforeImportPermissionSets(TempBlob, UpdateExistingPermissions, Handeld);
                    if not Handeld then
                        UpdateExistingPermissions := Confirm(UpdateExistingPermissionsLbl, true);
#if not CLEAN21
                    if IsImportNewVersion(InStream) then begin
                        ImportPermissionSets.SetSource(InStream);
                        ImportPermissionSets.SetUpdatePermissions(UpdateExistingPermissions);
                        ImportPermissionSets.Import();
                    end else begin
                        ImportTenantPermissionSets.SetSource(InStream);
                        ImportTenantPermissionSets.SetUpdatePermissions(UpdateExistingPermissions);
                        ImportTenantPermissionSets.Import();
```

The event should include the following parameters.
```
    [IntegrationEvent(false, false)]
    local procedure OnBeforeImportPermissionSets(TempBlob: Codeunit "Temp Blob"; var UpdateExistingPermissions: Boolean; var Handeld: Boolean)
    begin
    end;
```

Our user loves this function - so perhaps you can ask to include this function in standard instead of a new event...
```
    local procedure CheckImportFileForImportPermissionSets(TempBlob: Codeunit "Temp Blob")
    var
        AppID: Text;
        BeginText: Label 'PermissionSet AppID="';
        BeginTextPos: Integer;
        DeleteExistingUserGroupPermissionSets: Label 'The existing permission sets are updated. Should the existing access rights of the permission sets to be imported also be deleted before the import?';
        EndText: Label '" RoleName="';
        EndTextPos: Integer;
        Found: Boolean;
        InStr: InStream;
        MiddleText: Label '" RoleID="';
        MiddleTextPos: Integer;
        NullGuid: Guid;
        Number: Integer;
        RoleID: Text;
        TenantPermission: Record "Tenant Permission";
        TextTemp: Text;
        TextWork: Text;
        UserQuestionPerformed: Boolean;
    begin
        if not TempBlob.HasValue() then
            exit;

        TextWork := '';
        TempBlob.CreateInStream(InStr);
        while not (InStr.EOS) do begin
            Number := InStr.ReadText(TextTemp);
            TextWork := TextWork + TextTemp;
            EndTextPos := StrPos(TextWork, EndText);
            if EndTextPos <> 0 then begin
                MiddleTextPos := StrPos(TextWork, MiddleText);
                if MiddleTextPos <> 0 then begin
                    BeginTextPos := StrPos(TextWork, BeginText);
                    if BeginTextPos <> 0 then begin
                        AppID := CopyStr(TextWork, BeginTextPos + StrLen((BeginText)), MiddleTextPos - BeginTextPos - StrLen(BeginText));
                        RoleID := CopyStr(TextWork, MiddleTextPos + StrLen((MiddleText)), EndTextPos - MiddleTextPos - StrLen(MiddleText));
                        if StrLen(RoleID) <= MaxStrLen(TenantPermission."Role ID") then begin
                            TenantPermission.SetCurrentKey("App ID", "Role ID", "Object Type", "Object ID");
                            TenantPermission.SetRange("App ID", UpperCase(AppID));
                            TenantPermission.SetRange("Role ID", UpperCase(RoleID));
                            Found := not TenantPermission.IsEmpty;
                            if not Found then
                                if UpperCase(AppID) <> NullGuid then begin
                                    TenantPermission.SetRange("App ID", NullGuid);
                                    Found := not TenantPermission.IsEmpty;
                                end;
                            if Found then begin
                                if not UserQuestionPerformed then begin
                                    UserQuestionPerformed := true;
                                    if not CONFIRM(false, DeleteExistingUserGroupPermissionSets) then
                                        exit;
                                end;
                                TenantPermission.DeleteAll();
                            end;
                        end;
                    end;
                end;
                TextWork := '';
            end;
        end;
    end;

    [EventSubscriber(ObjectType::Page, Page::"Permission Sets", 'IAX_OnBeforeImportPermissionSets', '', true, true)]
    local procedure CheckImportFileForImportPermissionSets_OnBeforeImportPermissionSets(TempBlob: Codeunit "Temp Blob"; var UpdateExistingPermissions: Boolean; var Handeld: Boolean)
    var
    begin
        CheckImportFileForImportPermissionSets(TempBlob);
        UpdateExistingPermissions := true;
        Handeld := true
    end;

```

**Comment by pri-kise** [comment]
@JesperSchulz @miljance can this event request be avoided by your new PR?
https://github.com/microsoft/BCApps/pull/3184

**Comment by Martin-Wagner-AX** [comment]
Hello @miljance, hello @JesperSchulz, 
yes, with this event, I can integrate my logic and delete existing permissions before inserting the new definition. 
I want to delete permissions only after user confirmation. 
To avoid two confirm dialogs to the user, I would like to have the event in the page action as described, so that I can also bypass the standard confirmation.
But nevertheless, that's an improvement.
Kind regards
Martin

**Comment by miljance** [comment]
I think I understand your requirement now. Your goal seems to be avoiding two user dialogs. If that is the case, you indeed need a Handled pattern, but you would then just need to do the following in code:


```
                    begin
                        if not UploadIntoStream('Import', '', '', FileName, InStream) then
                            exit;

                        TempBlob.CreateOutStream(OutStream);
                        CopyStream(OutStream, InStream);

                        TempBlob.CreateInStream(InStream);
                        UpdateExistingPermissions := Confirm(UpdateExistingPermissionsLbl, true);
                        ImportPermissionSets.SetSource(InStream);
                        ImportPermissionSets.SetUpdatePermissions(UpdateExistingPermissions);
                        ImportPermissionSets.Import();

                        PermissionSetBuffer := Rec;
                        Rec.FillRecordBuffer();
                        if Rec.Get(PermissionSetBuffer.Type, PermissionSetBuffer."Role ID") then;
                    end;

```

add Event:

```
                    begin
                        if not UploadIntoStream('Import', '', '', FileName, InStream) then
                            exit;

                        TempBlob.CreateOutStream(OutStream);
                        CopyStream(OutStream, InStream);

                        TempBlob.CreateInStream(InStream);
                        OnBeforeImportPermissionSets(UpdateExistingPermissions, Handled);
                        if not Handled then begin
                            UpdateExistingPermissions := Confirm(UpdateExistingPermissionsLbl, true);
                            ImportPermissionSets.SetSource(InStream);
                            ImportPermissionSets.SetUpdatePermissions(UpdateExistingPermissions);
                            ImportPermissionSets.Import();
                        end;

                        PermissionSetBuffer := Rec;
                        Rec.FillRecordBuffer();
                        if Rec.Get(PermissionSetBuffer.Type, PermissionSetBuffer."Role ID") then;
                    end;
```

and in your Subscriber with the changes that I made you should just do the following:

```
                            ImportPermissionSets.SetSource(InStream);
                            ImportPermissionSets.SetUpdatePermissions(false);
                            ImportPermissionSets.Import();
```

Because calling the Impor XMLPort with a false parameter makes sure that all of the permissions are deleted from a permission set.
This looks a lot simpler than what you had so far in the function.

Event simpler, if you cannot wait for the Event you can hide standard action and create a new one using the three lines of code that I have provided for the import purposes.

**Comment by Martin-Wagner-AX** [comment]
Sorry - I do not agree...
In BC26 in xmlport 9864 "Import Permission Sets" we do have only this code and nowhere permissions will be deleted.

```
    var
        UpdatePermissions: Boolean;

    trigger OnPreXmlPort()
    begin
        DisallowEditingPermissionSetsForNonAdminUsers();
        OnAfterOnPreXmlPort(UpdatePermissions);
    end;

    procedure SetUpdatePermissions(NewUpdatePermissions: Boolean)
    begin
        UpdatePermissions := NewUpdatePermissions;
    end;

    [IntegrationEvent(false, false)]
    local procedure OnAfterOnPreXmlPort(var UpdatePermissions: Boolean)
    begin
    end;
```

For example:
I have a permission set call TEST including rights for table 3, 4 and 5. 
A user exports this permission set, deletes all line for table 4 and imports the file again.
With actual xmlport in BC26 permissions for table 4 won´t be deleted regardless of whether I answer the confirm with yes or no.

And if I subscribe to OnAfterOnPreXmlPort I do not know which permission sets are included in file, so I do not know which permission sets I should delete for my own.

**Comment by miljance** [comment]
The code is coming in BC26.1:
https://github.com/microsoft/BCApps/pull/3355
And yes, it supports deletion of permission prior to importing permission set.

**Comment by stkillen** [comment]
Sorry, this one slipped through the cracks. 

As pointed out @miljance added this functionality in 26.1 so closing this issue. 

For future requests to changes in System, follow the guidelines on https://github.com/microsoft/BCApps/blob/main/CONTRIBUTING.md

**Comment by JesperSchulz** [comment]
Thanks for reporting this. We agree, and we’ll publish a fix asap, either in an update for the current version or in the next major release. Please do not reply to this, as we do not monitor closed issues. If you have follow-up questions or requests, please create a new issue where you reference this one.

Build ID: <None>.
